--- CPieGraphDraw.php.orig      2014-02-12 10:25:36.912261759 +0400
+++ CPieGraphDraw.php   2014-02-12 10:29:38.063988265 +0400
@@ -286,14 +286,37 @@
                $max_host_len = 0;
                $max_name_len = 0;

-               for ($i = 0; $i < $this->num; $i++) {
-                       if (zbx_strlen($this->items[$i]['hostname']) > $max_host_len) {
-                               $max_host_len = zbx_strlen($this->items[$i]['hostname']);
-                       }
-                       if (zbx_strlen($this->items[$i]['name_expanded']) > $max_name_len) {
-                               $max_name_len = zbx_strlen($this->items[$i]['name_expanded']);
-                       }
-               }
+        for ($i = 0; $i < $this->num; $i++) {
+                if (zbx_strlen($this->items[$i]['host']) > $max_host_len) {
+                        $max_host_len = zbx_strlen($this->items[$i]['host']);
+                }
+                if (zbx_strlen($this->items[$i]['name']) > $max_name_len) {
+                        $max_name_len = zbx_strlen($this->items[$i]['name']);
+                }
+        }
+
+        $hide_hostname = true;
+        $leftXShift = 16;
+        $legend = new CImageTextTable($this->im, $leftXShift+10, $this->sizeY+$shiftY+5);
+        $legend->color = $this->getColor($this->graphtheme['textcolor'], 0);
+        $legend->rowheight = 14;
+        $legend->fontsize = 8;
+
+        for($i=1;$i<$this->num;$i++){
+            if ($this->items[$i-1]['hostid'] != $this->items[$i]['hostid']) {
+                $hide_hostname = false;
+                break;
+            }
+        }
+
+        $legend->addRow(array(
+            array('text'=>''),
+            array('text'=>''),
+            array('text'=>'value', 'align'=> 1, 'fontsize' => 9),
+            array('text'=>'% of total', 'align'=> 1, 'fontsize' => 9)
+        ));
+        $rowNum = 1;
+

                for ($i = 0; $i < $this->num; $i++) {
                        $color = $this->getColor($this->items[$i]['color'], 0);
@@ -323,47 +346,43 @@

                        $proc = $this->sum == 0 ? 0 : ($datavalue * 100) / $this->sum;

-                       if (isset($data) && isset($datavalue)) {
-                               $strvalue = sprintf(_('Value').': %s ('.(round($proc) != round($proc, 2) ? '%0.2f' : '%0.0f').'%%)',
-                                       convert_units(array(
-                                               'value' => $datavalue,
-                                               'units' => $this->items[$i]['units']
-                                       )),
-                                       $proc
-                               );
+                       if ($hide_hostname)
+                               $item_caption = str_pad($this->items[$i]['name'],$max_desc_len,' ');
+                       else
+                               $item_caption = str_pad($this->items[$i]['host'],$max_host_len,' ').': '.str_pad($this->items[$i]['name'],$max_desc_len,' ');

-                               $str = sprintf('%s: %s [%s] ',
-                                       str_pad($this->items[$i]['hostname'], $max_host_len, ' '),
-                                       str_pad($this->items[$i]['name_expanded'], $max_name_len, ' '),
-                                       $fncRealName
-                               );
+
+                       if (isset($data) && isset($datavalue)) {
+                               $legend->addCell($rowNum, array('text' => $item_caption));
+                               $legend->addCell($rowNum, array('text' => '['.$fncRealName.']'));
+                               $legend->addCell($rowNum, array('text' => convert_units(array('value' => $datavalue, 'units' => $this->items[$i]['units'], 'align'=> 2))));
+                               $legend->addCell($rowNum, array('text' => sprintf('%0.2f %%',$proc), 'align'=> 2));
                        }
                        else {
-                               $strvalue = sprintf(_('Value: no data'));
-                               $str = sprintf('%s: %s [ '._('no data').' ]',
-                                       str_pad($this->items[$i]['hostname'], $max_host_len, ' '),
-                                       str_pad($this->items[$i]['name_expanded'], $max_name_len, ' ')
-                               );
+                               $legend->addCell($rowNum, array('text' => $item_caption));
+                               $legend->addCell($rowNum, array('text' => '[ '. S_NO_DATA_SMALL . ' ]'));
                        }

                        imagefilledrectangle(
                                $this->im,
                                $this->shiftXleft,
-                               $this->sizeY + $shiftY + 14 * $i - 5,
+                               $this->sizeY + $shiftY + 14 * $rowNum - 5,
                                $this->shiftXleft + 10,
-                               $this->sizeY + $shiftY + 5 + 14 * $i,
+                               $this->sizeY + $shiftY + 5 + 14 * $rowNum,
                                $color
                        );

                        imagerectangle(
                                $this->im,
                                $this->shiftXleft,
-                               $this->sizeY + $shiftY + 14 * $i - 5,
+                               $this->sizeY + $shiftY + 14 * $rowNum - 5,
                                $this->shiftXleft + 10,
-                               $this->sizeY + $shiftY + 5 + 14 * $i,
+                               $this->sizeY + $shiftY + 5 + 14 * $rowNum,
                                $this->getColor('Black No Alpha')
                        );

+                       $rowNum++;
+/*
                        imageText(
                                $this->im,
                                8,
@@ -403,11 +422,15 @@
                                $this->getColor($this->graphtheme['textcolor'], 0),
                                $strvalue
                        );
+*/
                }

+/*
                if ($this->sizeY < 120) {
                        return;
                }
+*/
+               $legend->draw();
        }

        protected function drawElementPie($values) {
@@ -443,7 +466,7 @@
                        $sizeY = (int) $sizeY * 0.95;
                }

-               $xc = $x = (int) $this->sizeX / 2 + $this->shiftXleft;
+               $xc = $x = (int) $this->fullSizeX / 2;
                $yc = $y = (int) $this->sizeY / 2 + $this->shiftY;

                $anglestart = 0;
@@ -518,7 +541,7 @@

                list($sizeX, $sizeY) = $this->calc3DAngle($sizeX, $sizeY);

-               $xc = $x = (int) $this->sizeX / 2 + $this->shiftXleft;
+               $xc = $x = (int) $this->fullSizeX / 2;
                $yc = $y = (int) $this->sizeY / 2 + $this->shiftY;

                // bottom angle line
@@ -650,8 +673,8 @@
                }

                if ($this->drawLegend == 1) {
-                       $this->sizeX -= $this->shiftXleft + $this->shiftXright + $this->shiftlegendright;
-                       $this->sizeY -= $this->shiftY + $this->shiftYLegend + 12 * $this->num + 8;
+                       $this->sizeX -= ($this->shiftXleft * 2);
+                       $this->sizeY -= ($this->shiftY+$this->shiftYLegend + 14 * ($this->num + 1) + 8);
                }
                else {
                        $this->sizeX -= $this->shiftXleft * 2;
